### PR TITLE
Bulk: Refactors async task handling and reduces allocations

### PR DESCRIPTION
**Pull request by ealsur**
Based on recent investigations, some improvement opportunities were identified on the Bulk code.

## TaskCompletionSource

Our current use of TaskCompletionSource did not go along the recommendations: https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#always-create-taskcompletionsourcet-with-taskcreationoptionsruncontinuationsasynchronously and https://devblogs.microsoft.com/premier-developer/the-danger-of-taskcompletionsourcet-class/. Based on David's points, we should be using `TaskCreationOptions.RunContinuationsAsynchronously` to avoid executing the continuation inline and having the thread waiting on it. 

Changing to use `TaskCreationOptions.RunContinuationsAsynchronously` would make the continuation go back to the thread pool and acquire a new thread whenever it needs to continue, and help prevent thread-pool starvation.

## ContinueWith

`ContinueWith` was being used to handle Timer dispatch on `BatchAsyncStreamer`. Once the timer was up, it would dispatch the current batch. Following the same source of recommendations (https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-await-over-continuewith), we remove the ContinueWith in favor of `async/await`. Calling ContinueWith allocates another task per operation (it wraps your delegate in a task object) instead of re-using the state machine instance as the continuation, and also has to allocate a context that includes the callback + options. Using ContinueWith also prevents leveraging any optimizations that the framework includes in async operations.

## Performance comparison

Executing operations as concurrent Task, 1Kb item creates on a machine targeting a live Cosmos DB account. Each set is executed 5 times and the average is obtained.

### 100K RU provisioned collection

**Before**

Documents| Latency average (ms)
------ | ------
10 | 142
100 | 144
500 | 173
1000 | 229
2000 | 274
3000 | 386
5000 | 568
6000 | 676
10000 | 1058
50000 | 4734
100000 | 9346
200000 | 18456

**After**

Documents| Latency average (ms)
------ | ------
10 | 93
100 | 95
500 | 115
1000 | 158
2000 | 243
3000 | 333
5000 | 484
6000 | 566
10000 | 967
50000 | 4058
100000 | 8115
200000 | 17667

### 10K RU provisioned collection

**Before**

Documents| Latency average (ms)
------ | ------
10 | 132
100 | 141
500 | 192
1000 | 369
2000 | 800
3000 | 1314
5000 | 2070
6000 | 2610

**After**

Documents| Latency average (ms)
------ | ------
10 | 85
100 | 85
500 | 231
1000 | 351
2000 | 735
3000 | 1222
5000 | 2032
6000 | 2423

While the performance gain is not huge, it does show a consistent 5-10% improvement on repeated tests.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs [hunk 1]
```diff
@@ -190,15 +194,49 @@ private static void AddHeadersToRequestMessage(RequestMessage requestMessage, st
...
         private async Task<string> ResolvePartitionKeyRangeIdAsync(
             ItemBatchOperation operation,
             CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
-            PartitionKeyDefinition partitionKeyDefinition = await this.cosmosContainer.GetPartitionKeyDefinitionAsync(cancellationToken);
-            CollectionRoutingMap collectionRoutingMap = await this.cosmosContainer.GetRoutingMapAsync(cancellationToken);
+            if (!this.lazyExecutorCaches.ValueInitialized)
+            {
+                await this.lazyExecutorCaches.GetValueAsync(NoOpTrace.Singleton, cancellationToken);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:228]: Create a new root trace rather than a NoOpTrace. NoOpTrace will lose all diagnostics. A new root will at least capture retries and failures for the cache.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs [hunk 2]
```diff
@@ -190,15 +194,49 @@ private static void AddHeadersToRequestMessage(RequestMessage requestMessage, st
...
+            }
+
+            if (this.lazyExecutorCaches.Result.Failed)
+            {
+                throw this.lazyExecutorCaches.Result.Exception.InnerException;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:233]: Is the BatchAsyncContainerExecutor.cs re-used at all? By making them lazy if any of the operations hit a failure it will always return that original failure, and retrying will not work.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs [hunk 3]
```diff
@@ -190,15 +194,49 @@ private static void AddHeadersToRequestMessage(RequestMessage requestMessage, st
             }
         }
 
+        private async Task<TryCatch<ExecutorCaches>> TryInitializeExecutorCachesAsync(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:197]: Is it worth add a perf test for this scenario?

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:197]: Sure, I think it will always be good. I'm not sure with the current state of Diagnostics but it will probably be good.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs (hunk 2 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:233]: `BatchAsyncContainerExecutor` is reused, in fact, there is a `BatchAsyncContainerExecutorCache` that maintains one instance per Container to make sure users always get the same instance on different `GetContainer` calls.

Before this change we were using the Cached container properties, would that issue new requests if the previous one failed? 

Should we then not cache the exception and re-initialize the Lazy in that case?

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs (hunk 1 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:228]: Worth tracking when we fix the diagnostics in Bulk

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs (hunk 2 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs]: You are right, the AsyncCache would bring nothing but trouble and the benefit is not that great

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Bulk/MockedItemOfTBulkBenchmark.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos;
+
+    public class MockedItemOfTBulkBenchmark : IItemBulkBenchmark
+    {
+        public MockedItemBenchmarkHelper BenchmarkHelper { get; set; }
+
+        public async Task CreateItem()
+        {
+            ItemResponse<ToDoActivity> response = await this.BenchmarkHelper.TestContainer.CreateItemAsync<ToDoActivity>(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Bulk/MockedItemOfTBulkBenchmark.cs:18]: Shouldn't this be doing x amount of creates to avoid waiting on the time to send the single request?

Comment (ealsur) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Bulk/MockedItemOfTBulkBenchmark.cs:18]: Dispatch of the batch is both time bound and operation count bound, there is no way to assert and pack the batch to make the test always return the same result/behavior

Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Bulk/MockedItemOfTBulkBenchmark.cs:18]: Shouldn't it be the same if we always process 500 items? Even if there is 5 left over that require the timer to finish it.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs (hunk 3 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs]: Done

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Bulk/MockedItemOfTBulkBenchmark.cs (hunk 4 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Bulk/MockedItemOfTBulkBenchmark.cs:17]: Sure, but that then would be the same case for 1 item

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs [hunk 5]
```diff
@@ -192,13 +192,18 @@ private static void AddHeadersToRequestMessage(RequestMessage requestMessage, st
 
         private async Task<string> ResolvePartitionKeyRangeIdAsync(
             ItemBatchOperation operation,
+            ITrace trace,
             CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
-            PartitionKeyDefinition partitionKeyDefinition = await this.cosmosContainer.GetPartitionKeyDefinitionAsync(cancellationToken);
+            ContainerProperties cachedContainerPropertiesAsync = await this.cosmosContainer.GetCachedContainerPropertiesAsync(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:199]: What is the reason for this change?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Mocks/BatchResponsePayloadWriter.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+        public BatchResponsePayloadWriter(List<TransactionalBatchOperationResult> results)
+        {
+            this.results = results;
+        }
+
+        internal async Task<MemoryStream> GeneratePayloadAsync()
+        {
+            MemoryStream responseStream = new MemoryStream();
+            await responseStream.WriteRecordIOAsync(default, this.WriteOperationResult);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Mocks/BatchResponsePayloadWriter.cs:29]: Is there any reason we aren't writing this to a byte array and then just creating a new memory stream using the same byte array?

Comment (ealsur) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Mocks/BatchResponsePayloadWriter.cs:30]: Good point, done.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs (hunk 5 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:199]: Be able to eventually pass the Trace otherwise we are losing the information on the first request

**Review by j82w — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs (hunk 5 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:199]: That's a gap we need to fix. The GetPartitionKeyDefinitionAsync should take in ITrace.
